### ISSUE TITLE
Would apex still be useful for non Volta architectures?

### ISSUE THREAD

**Issue by im9uri** [NVIDIA/apex#14]
I was looking into the library, and it seems that the assumption is that the GPU is a Volta architecture.

[This link](https://github.com/u39kun/deep-learning-benchmark#benchmark-results) shows some benchmarks for fp16 training and inference, and the 1080 Ti doesn't gain that much performance from fp16.

Would it be useful to apply this library for GPUs besides Titan V and V100?

**Comment by mcarilli** [comment]
Hi Gyuri, 

Thank you for your interest in Apex!  Apex is not designed to offer performance gains over "pure FP16" training.  Apex is designed to help with numerical stability.  The intent of Apex is to provide numerical stability comparable to "pure FP32" while maintaining almost all the performance of "pure FP16" training.

Apex is intended primarily for Volta, although for Pascal, it may still be helpful.  If you're interested in details:

Pascals (like 1080Ti) support FP16 ops only as vector instructions.  This means a couple things:  First of all, at best you could only ever expect 2X speedup, whether compute bound or bandwidth bound.  Second, for gemms and convolutions,  any accumulate (ie the dot product of a row of matrix A with a column of matrix B) must be done in FP16 (unless you want to upconvert to FP32 while accumulating, which would kill your compute throughput).  Therefore, FP16 gemms and convolutions on Pascal must be (internally) either somewhat unstable, or comparable in performance to FP32.

Volta Tensor Cores, by contrast, are specialized hardware for doing gemms in tiles.  They perform arithmetic in FP16 and accumulate into FP32.  Therefore, gemms (and convolutions, which can be treated as implicit gemms) on FP16 data are numerically stable on Volta.  The danger points of FP16 networks on Volta tend to be 1.  parameter updates, 2.  gradient underflow, and 3.  any intermediate reduction ops (sums, norms) that might encounter a large value (> 65,504).   

Apex is intended to help with 1, 2, and 3.  FP16_Optimizer takes care of issues 1 and 2 automatically.  Amp takes care of 1, 2, and 3 automatically.  On Volta, this should be enough to make everything stable.

Bottom line:  1080Ti (Pascal) is a great chip for training in FP32, but for FP16 it has some issues that Apex can't help with (the convolutions and gemms).  1, 2, and 3 are also danger points on Pascal, so Apex may still be helpful.

**Comment by im9uri** [comment]
Thanks for the helpful information!

**Comment by gcp** [comment]
>Would it be useful to apply this library for GPUs besides Titan V and V100?

For completeness, it's useful also on the P100 (which is available with many cloud providers) as that also has full speed fp16, even though it's a Pascal based card.

**Comment by ngimel** [comment]
Also for completeness, pure fp16 training on P100 is not recommended and not supported, pseudo-fp16 (with storage in fp16, and math in fp32) can be used, but it's math throughput is approx equal to fp32 throughput on P100, so the only benefit is coming from less memory needed.

**Comment by gcp** [comment]
Why is it not recommended? Is it because the intermediate adds accumulate in fp16 instead of fp32 as on Volta?

What happens if you try to use apex on that configuration? Do you just get less accurate convolutions?

**Comment by ngimel** [comment]
Yes, because of accumulation in fp16. If you use Apex on p100, it would still call pytorch, and pytorch does not support fp16 accumulation, you'll get pseudo-fp16 whether you want it or not.
